Better behavior for #detect_model

mach-kernel opened the pull request:
It appears that this library services ActiveRecord primarily, so I wanted to use `#klass` instead of reifying it from the table name because it broke for me earlier in the following situation: 

- Table named `diy_links`.
- Model named `DIYLink`.
- Method attempts to reify `DiyLink` and fails.

I will add a spec later today, hope this is helpful! 👍

davidcelis commented:
@mach-kernel Thanks, this is great! Happy to merge this once you push the spec up 😄

mach-kernel commented:
@davidcelis I think this should cover it! I made a few other changes:

- Refactored `get_default_per_page_for_kaminari` and `default_per_page_for_will_paginate` to not suppress all exceptions (necessary, because otherwise the `NameError` we use in our spec to test against using `#klass` vs reification from table name would never percolate up!)
- Changed `#detect_model` to `#extract_per_page_from_model`. It finds the model then attempts to get whatever the correct `per_page` attribute is provided by the two aforementioned methods. I am not good with naming things, so if there is a better one we should use it :dancer: 
- Rails spec bug: a case (presumably around the catchalls in the first bullet) was fixed and another added. We never undefined the accessors from `Fixnum` while testing whether or not the class responded to those methods. I added that behavior to your spec and another to handle what happens when a `nil` value is yielded by one of those accessors. 

Thanks!

mach-kernel commented on spec/rails_spec.rb:
```diff
@@ -302,6 +302,26 @@ class Fixnum
         )
       end
+
+      # This spec has to be last because if we undefine these methods
```
Alternately, put it first so we don't have to invoke `undef_method` at all?

mach-kernel commented:
:wave: 

Sorry to bug, have you had a chance to look this over? :smile_cat:

mach-kernel commented on spec/active_record_spec.rb:
```diff
@@ -0,0 +1,49 @@
+
+    it 'correctly produces the correct model independent of table name' do 
+      expect { ApiPagination.paginate(collection) }.not_to raise_error
+    end
```
This is a tough one given that the test is sort of contrived to begin with / I didn't label it properly. Swapping out the table name for something that would fail string reification smells a bit weak as assertion, but since the library does not transform the individual objects in the collection I didn't know what else to do. I am going to add extra specs around making sure the correct methods are invoked as part of paginating and make the `it` clearer.

davidcelis commented on spec/active_record_spec.rb:
```diff
@@ -0,0 +1,54 @@
+
+  if ApiPagination.config.paginator == :kaminari
+    context 'pagination with kaminari' do
+      before { ApiPagination.config.paginator = :kaminari }
```
We should be able to get rid of this `before` proc since the paginator is already `:kaminari`

davidcelis commented on spec/active_record_spec.rb:
```diff
@@ -0,0 +1,54 @@
+    require 'will_paginate/active_record'
+  
+    context 'pagination with will_paginate' do 
+      before { ApiPagination.config.paginator = :will_paginate }
```
We can get rid of this one too 😄

davidcelis commented on spec/rails_spec.rb:
```diff
@@ -271,13 +279,25 @@ class Fixnum
 
           expect(response.header['Per-Page']).to eq(
             case ApiPagination.config.paginator
-            when :pagy          then Pagy::VARS[:items].to_s
```
You'll need to add this back in (and also add it to your new spec below!), probably got lost in a merge conflict!

mach-kernel commented:
Eep, thanks for looking this over again 🦄

davidcelis commented on spec/rails_spec.rb:
```diff
@@ -279,5 +287,18 @@ class Fixnum
+        get :index_with_no_per_page, params: {count: 100}
+
+        expect(response.header['Per-Page']).to eq(
+          case ApiPagination.config.paginator
```
Would you add the `when :pagy` case into this new test too?

davidcelis commented:
It looks like the tests are failing with `undefined method `to_sql' for [#<ActiveRecord::Relation []>, nil]:Array`, probably because of an implementation detail I caused to make Pagy work easily; you can fix this by replacing it with `.first.to_sql` 😅

mach-kernel commented:
@davidcelis arg, looks like a few other bugs too. I will take a look at these at some point today. We will get this working. 😂

mach-kernel commented:
One failure in particular is hard to track down. It's not being triggered by Pagy, but rather, by the `.count` in the Pagy constructor. We're going to go down the rabbit hole now...

```
3) ActiveRecord Support reflections does not fail if table name is not snake cased class name
   Failure/Error: expect { ApiPagination.paginate(collection) }.to_not raise_error

     expected no Exception, got #<TypeError: no implicit conversion of String into Integer> with backtrace:
       # ./lib/api-pagination.rb:67:in `pagy_from'
       # ./lib/api-pagination.rb:56:in `paginate_with_pagy'
       # ./lib/api-pagination.rb:13:in `paginate'
       # ./spec/active_record_spec.rb:53:in `block (4 levels) in <top (required)>'
       # ./spec/active_record_spec.rb:53:in `block (3 levels) in <top (required)>'
   # ./spec/active_record_spec.rb:53:in `block (3 levels) in <top (required)>'
```

Looks like `nulldb` doesn't seem to be decorating type data from the schema onto the model classes. It's causing a failure inside the `active_record/relations/calculations`, right around here, inside `#execute_simple_calculation`:

```ruby
type   = result.column_types.fetch(column_alias) do
  type_for(column_name)
end
```

`nulldb` does implement this API inside [EmptyResult](https://github.com/nulldb/nulldb/blob/master/lib/active_record/connection_adapters/nulldb_adapter/empty_result.rb), so what gives? Breaking in here at runtime gives us this:

```ruby
{"id"=>
  #<ActiveRecord::ConnectionAdapters::NullDBAdapter::Column:0x007f9647819338
   @collation=nil,
   @comment=nil,
   @default=nil,
   @default_function=nil,
   @name="id",
   @null=false,
   @sql_type_metadata=
    #<ActiveRecord::ConnectionAdapters::SqlTypeMetadata:0x007f96478193d8
     @limit=nil,
     @precision=nil,
     @scale=nil,
     @sql_type=:primary_key,
     @type=nil>,
   @table_name=nil>,
 "foo"=>
  #<ActiveRecord::ConnectionAdapters::NullDBAdapter::Column:0x007f9647819220
   @collation=nil,
   @comment=nil,
   @default=nil,
   @default_function=nil,
   @name="foo",
   @null=true,
   @sql_type_metadata=
    #<ActiveRecord::ConnectionAdapters::SqlTypeMetadata:0x007f9647819298
     @limit=nil,
     @precision=nil,
     @scale=nil,
     @sql_type=:string,
     @type=nil>,
   @table_name=nil>}
```

Notice the empty `type` fields on `SqlTypeMetadata`. This takes us on a small adventure into the `nulldb` adapter, namely here: 

```ruby
def args_with_optional_cast_type(col_def)
  default_column_arguments(col_def).tap do |args|
    if defined?(ActiveRecord::ConnectionAdapters::SqlTypeMetadata)
      meta = ActiveRecord::ConnectionAdapters::SqlTypeMetadata.new(sql_type: col_def.type)
      args.insert(2, meta_with_limit!(meta, col_def))
    elsif initialize_column_with_cast_type?
      args.insert(2, meta_with_limit!(lookup_cast_type(col_def.type), col_def))
    else
      args[2] = args[2].to_s + "(#{col_def.limit})" if col_def.limit
    end
  end
end
```

Maybe we can add the `type` attribute back:

```ruby
meta = ActiveRecord::ConnectionAdapters::SqlTypeMetadata.new(sql_type: col_def.type)
```

OK, now we're getting type attributes, but WTF, still nothing in `#column_types`:

```ruby
[1] pry(#<Foo::ActiveRecord_Relation>)> @klass.columns_hash
=> {"id"=>
  #<ActiveRecord::ConnectionAdapters::NullDBAdapter::Column:0x007ff3b524bf20
   @collation=nil,
   @comment=nil,
   @default=nil,
   @default_function=nil,
   @name="id",
   @null=false,
   @sql_type_metadata=
    #<ActiveRecord::ConnectionAdapters::SqlTypeMetadata:0x007ff3b5250480
     @limit=nil,
     @precision=nil,
     @scale=nil,
     @sql_type=:integer,
     @type=:integer>,
   @table_name=nil>,
 "foo"=>
  #<ActiveRecord::ConnectionAdapters::NullDBAdapter::Column:0x007ff3b524be08
   @collation=nil,
   @comment=nil,
   @default=nil,
   @default_function=nil,
   @name="foo",
   @null=true,
   @sql_type_metadata=
    #<ActiveRecord::ConnectionAdapters::SqlTypeMetadata:0x007ff3b524be80
     @limit=nil,
     @precision=nil,
     @scale=nil,
     @sql_type=:text,
     @type=:text>,
   @table_name=nil>}
[2] pry(#<Foo::ActiveRecord_Relation>)> result.column_types
=> []
```

Something is still wrong...let's take a look at the column metadata class from `nulldb`: 

```ruby
class ActiveRecord::ConnectionAdapters::NullDBAdapter
  class Column < ::ActiveRecord::ConnectionAdapters::Column

    private

    def simplified_type(field_type)
      super || simplified_type_from_sql_type
    end

    def simplified_type_from_sql_type
      case sql_type
      when :primary_key
        :integer
      when :string
        :string
      end
    end

  end
end
```

A fair bit of Googling later we find that [#simplified_type](https://github.com/rails/rails/pull/15158/files) was replaced in Rails 4.2. There is a new [type registry API](http://api.rubyonrails.org/classes/ActiveRecord/Type.html) that needs to be implemented on the `nulldb` side (i.e. by providing type definitions for that adapter) in order to remedy this failure, in addition to the aforementioned param fix. I'll check back in after I sort this out.

mach-kernel commented:
@davidcelis `nulldb` changes were merged and specs updated. It's ready for another round of review!

mach-kernel commented:
@davidcelis :wave: heya, just wanted to ping about this one before I forget. LMK if I can be helpful. 🦄

davidcelis commented:
Oops, sorry this fell off my radar! Thanks again, @mach-kernel!
